Temporary file is deleted before a controller returns a response

### Play Version

3.0.4

### API

Scala

### Operating System

macOS 14.4.1

### JDK

```
openjdk version "21.0.2" 2024-01-16 LTS
OpenJDK Runtime Environment Corretto-21.0.2.13.1 (build 21.0.2+13-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.2.13.1 (build 21.0.2+13-LTS, mixed mode, sharing)
```

### Library Dependencies

None

### Expected Behavior

I have prepared the following Controller method.

```scala
  def fileSize() =
    Action.async(parse.multipartFormData) { implicit request =>
      request.body.file("file") match {
        case Some(image) =>
          for {
            path <- Future(image.ref.path)
            _ <- Future(System.gc())
            _ <- Future(Thread.sleep(100))
          } yield Ok(Files.size(path).toString)
        case None =>
          Future.successful(BadRequest("No image supplied"))
      }
    }
```

The endpoint is as follows.

```
POST    /fileSize                   controllers.HomeController.fileSize()
```

<!-- Please describe the expected behavior of the issue, starting from the first action. -->

1. Send a POST request.
    - `echo -n Hello > /tmp/a; curl -X POST http://localhost:9000/fileSize -F"file=@/tmp/a"`
2. The server returns the file size and the command displays 5.

### Actual Behavior


1. Send a POST request.
    - `echo -n Hello > /tmp/a; curl -X POST http://localhost:9000/fileSize -F"file=@/tmp/a"`
2. The server returns an error page.
3. The server logs following

```
2024-06-18 18:25:01 ERROR p.api.http.DefaultHttpErrorHandler  

! @85ph4o99f - Internal server error, for (POST) [/fileSize] ->
 
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[NoSuchFileException: /var/folders/r2/qjlxkmsn49v6mtmb6dgs3xrc0000gr/T/playtemp8649900777384842915/multipartBody4548848415921731547asTemporaryFile]]
	at play.api.http.HttpErrorHandlerExceptions$.$anonfun$convertToPlayException$2(HttpErrorHandler.scala:400)
	at scala.Option.map(Option.scala:242)
	at play.api.http.HttpErrorHandlerExceptions$.convertToPlayException(HttpErrorHandler.scala:398)
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:390)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:267)
	at play.core.server.PekkoHttpServer$$anonfun$invokeAction$1$1.applyOrElse(PekkoHttpServer.scala:482)
	at play.core.server.PekkoHttpServer$$anonfun$invokeAction$1$1.applyOrElse(PekkoHttpServer.scala:474)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490)
	at org.apache.pekko.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:73)
	at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:110)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
	at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:110)
	at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
	at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.nio.file.NoSuchFileException: /var/folders/r2/qjlxkmsn49v6mtmb6dgs3xrc0000gr/T/playtemp8649900777384842915/multipartBody4548848415921731547asTemporaryFile
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:171)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1853)
	at java.base/java.nio.file.Files.size(Files.java:2462)
	at controllers.HomeController.$anonfun$fileSize$7(HomeController.scala:36)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	... 5 common frames omitted
```

The file pointed to by TemporaryFile seems to be erased during GC, but I suppose that it would not be GC'd until it returns a response.

### Reproducible Test Case

https://github.com/tkawachi/play-tmpfile-deleted/commit/c9dbcd7c00863afc5b2f070cdc7fe9ceabe85dda

--- a/core/play-integration-test/src/test/scala/play/it/action/FormActionSpec.scala
+++ b/core/play-integration-test/src/test/scala/play/it/action/FormActionSpec.scala
@@ -4,6 +4,10 @@
 
 package play.it.action
 
+import java.nio.file.Files
+
+import scala.concurrent.Future
+
 import org.apache.pekko.actor.ActorSystem
 import org.apache.pekko.stream.Materializer
 import play.api._
@@ -11,7 +15,9 @@ import play.api.data._
 import play.api.data.format.Formats._
 import play.api.data.Forms._
 import play.api.libs.Files.TemporaryFile
+import play.api.libs.Files.TemporaryFileCreator
 import play.api.mvc.MultipartFormData
+import play.api.mvc.MultipartFormData.FilePart
 import play.api.mvc.Results._
 import play.api.routing.Router
 import play.api.test.FakeRequest
@@ -64,6 +70,22 @@ class FormActionSpec extends PlaySpecification with WsTestClient {
               val user = userForm.bindFromRequest().get
               Ok(s"${user.name} - ${user.email}")
           }
+        case SirdPost(p"/tmpfileexists") =>
+          // return true if the temporary file exists.
+          // this is used to test if the temporary file is kept while processing the request.
+          defaultActionBuilder(playBodyParsers.multipartFormData).async { implicit request =>
+            request.body.file("file") match {
+              case Some(file) =>
+                for {
+                  path <- Future(file.ref.path)
+                  // make sure the temporary file is kept even if GC is called
+                  _ <- Future(System.gc())
+                  _ <- Future(Thread.sleep(100))
+                } yield Ok(if (Files.exists(path)) "exists" else "not exists")
+              case None =>
+                Future.successful(BadRequest("No file"))
+            }
+          }
       }
     }.application
   }
@@ -98,6 +120,22 @@ class FormActionSpec extends PlaySpecification with WsTestClient {
         override def running() = {
           val request = FakeRequest(POST, "/multipart/wrapped-max-length").withMultipartFormDataBody(multipartBody)
           contentAsString(route(app, request).get) must beEqualTo("Player - play@email.com")
+        }
+      }
+
+      "keep TemporaryFiles of multipart request while processing" in new WithApplication(application) {
+        override def running(): Unit = {
+          val tempFileCreator = app.injector.instanceOf[TemporaryFileCreator]
+          val tempFile        = tempFileCreator.create()
+          Files.writeString(tempFile.path, "Hello")
+
+          val multipartBody = MultipartFormData[TemporaryFile](
+            dataParts = Map.empty,
+            files = Seq(FilePart("file", "file.txt", Some("text/plain"), tempFile)),
+            badParts = Seq.empty
+          )
+          val request = FakeRequest(POST, "/tmpfileexists").withMultipartFormDataBody(multipartBody)
+          contentAsString(route(app, request).get) must beEqualTo("exists")
         }
       }
     }

--- a/core/play/src/main/scala/play/api/mvc/Action.scala
+++ b/core/play/src/main/scala/play/api/mvc/Action.scala
@@ -432,6 +432,12 @@ trait ActionBuilder[+R[_], B] extends ActionFunction[Request, R] {
       def apply(request: Request[A]) =
         try {
           invokeBlock(request, block)
+            .andThen { _ =>
+              // This andThen block is used to keep a reference to the request until invokeBlock() is completed.
+              // If the request contains TemporaryFiles, they will be deleted when they are garbage collected.
+              // By keeping a reference to the request, it prevents the TemporaryFiles from becoming GC targets.
+              request
+            }(self.executionContext)
         } catch {
           // NotImplementedError is not caught by NonFatal, wrap it
           case e: NotImplementedError => throw new RuntimeException(e)

--- a/core/play/src/main/scala/play/api/mvc/Action.scala
+++ b/core/play/src/main/scala/play/api/mvc/Action.scala
@@ -437,7 +437,7 @@ trait ActionBuilder[+R[_], B] extends ActionFunction[Request, R] {
               // If the request contains TemporaryFiles, they will be deleted when they are garbage collected.
               // By keeping a reference to the request, it prevents the TemporaryFiles from becoming GC targets.
               request
-            }(self.executionContext)
+            }(Execution.trampoline)
         } catch {
           // NotImplementedError is not caught by NonFatal, wrap it
           case e: NotImplementedError => throw new RuntimeException(e)

--- a/core/play/src/main/scala/play/core/j/JavaAction.scala
+++ b/core/play/src/main/scala/play/core/j/JavaAction.scala
@@ -124,7 +124,15 @@ abstract class JavaAction(val handlerComponents: JavaHandlerComponents)
           .parseBody(
             parser,
             request.asScala(),
-            (r: Request[?]) => invocation(r.asJava).toCompletableFuture.asScala.map(_.asScala())
+            (r: Request[?]) =>
+              invocation(r.asJava).toCompletableFuture.asScala
+                .map(_.asScala())
+                .andThen { _ =>
+                  // This andThen block is used to keep a reference to the request until invocation() is completed.
+                  // If the request contains TemporaryFiles, they will be deleted when they are garbage collected.
+                  // By keeping a reference to the request, it prevents the TemporaryFiles from becoming GC targets.
+                  r
+                }(trampoline)
           )(executionContext)
           .map(_.asJava)
           .asJava

--- a/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/main/java/controller/HomeController.java
+++ b/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/main/java/controller/HomeController.java
@@ -10,6 +10,9 @@ import play.mvc.Http;
 import play.mvc.Result;
 
 import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
 
 public class HomeController extends Controller {
 
@@ -37,6 +40,35 @@ public class HomeController extends Controller {
                         + "contents: "
                         + contents
                         + "\n");
+    }
+
+    public CompletionStage<Result> multipartFormUploadTmpFileExists(Http.Request request) throws IOException {
+        return CompletableFuture.supplyAsync(() -> request
+                        .body()
+                        .<Files.TemporaryFile>asMultipartFormData()
+                        .getFile("file").getRef().path())
+                .thenApplyAsync(
+                        path -> {
+                            System.gc();
+                            return path;
+                        })
+                .thenApplyAsync(
+                        path -> {
+                            try {
+                                TimeUnit.MILLISECONDS.sleep(100);
+                            } catch (InterruptedException e) {
+                                throw new RuntimeException(e);
+                            }
+                            return path;
+                        })
+                .thenApplyAsync(
+                        path -> {
+                            if (java.nio.file.Files.exists(path)) {
+                                return ok("exists");
+                            } else {
+                                return ok("not exists");
+                            }
+                        });
     }
 
 }

--- a/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/main/java/controller/HomeController.java
+++ b/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/main/java/controller/HomeController.java
@@ -42,6 +42,8 @@ public class HomeController extends Controller {
                         + "\n");
     }
 
+    private java.nio.file.Path tmpPath = null; // not best practice, but for testing it's good enough
+
     public CompletionStage<Result> multipartFormUploadTmpFileExists(Http.Request request) throws IOException {
         return CompletableFuture.supplyAsync(() -> request
                         .body()
@@ -50,6 +52,7 @@ public class HomeController extends Controller {
                 .thenApplyAsync(
                         path -> {
                             System.gc();
+                            tmpPath = path;
                             return path;
                         })
                 .thenApplyAsync(
@@ -71,4 +74,22 @@ public class HomeController extends Controller {
                         });
     }
 
+    public Result gc(Http.Request request) {
+        System.gc();
+        try {
+            // Give garbage collector some time
+            TimeUnit.MILLISECONDS.sleep(100);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+        return ok();
+    }
+
+    public Result checkTmpFileStillExists(Http.Request request) {
+        if (java.nio.file.Files.exists(tmpPath)) {
+            return ok("exists");
+        } else {
+            return ok("not exists");
+        }
+    }
 }

--- a/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/main/resources/routes
+++ b/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/main/resources/routes
@@ -2,3 +2,4 @@
 
 POST     /multipart-form-data-no-files             controllers.HomeController.multipartFormUploadNoFiles(request: Request)
 POST     /multipart-form-data                      controllers.HomeController.multipartFormUpload(request: Request)
+POST     /multipart-form-data-tmpfileexists        controllers.HomeController.multipartFormUploadTmpFileExists(request: Request)

--- a/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/main/resources/routes
+++ b/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/main/resources/routes
@@ -3,3 +3,5 @@
 POST     /multipart-form-data-no-files             controllers.HomeController.multipartFormUploadNoFiles(request: Request)
 POST     /multipart-form-data                      controllers.HomeController.multipartFormUpload(request: Request)
 POST     /multipart-form-data-tmpfileexists        controllers.HomeController.multipartFormUploadTmpFileExists(request: Request)
+GET      /gc                                       controllers.HomeController.gc(request: Request)
+GET      /check-tmp-file-still-exists              controllers.HomeController.checkTmpFileStillExists(request: Request)

--- a/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/test/java/controllers/HomeControllerTest.java
+++ b/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/test/java/controllers/HomeControllerTest.java
@@ -22,6 +22,7 @@ import play.test.WithApplication;
 import static java.nio.file.Files.write;
 import static org.junit.Assert.assertEquals;
 import static play.mvc.Http.Status.OK;
+import static play.test.Helpers.GET;
 import static play.test.Helpers.POST;
 import static play.test.Helpers.route;
 
@@ -102,6 +103,30 @@ public class HomeControllerTest extends WithApplication {
         String content = result.body().consumeData(mat).thenApply(bs -> bs.utf8String()).toCompletableFuture().get(5, TimeUnit.SECONDS);
         assertEquals(OK, result.status());
         assertEquals("exists", content);
+
+        // Now let's check if the tmp file still gets removed when garbage collection takes place
+
+        request = new Http.RequestBuilder()
+                .method(GET)
+                .uri("/check-tmp-file-still-exists");
+        result = route(app, request);
+        content = result.body().consumeData(mat).thenApply(bs -> bs.utf8String()).toCompletableFuture().get(5, TimeUnit.SECONDS);
+        assertEquals(OK, result.status());
+        assertEquals("exists", content);
+
+        request = new Http.RequestBuilder()
+                .method(GET)
+                .uri("/gc");
+        result = route(app, request);
+        assertEquals(OK, result.status());
+
+        request = new Http.RequestBuilder()
+                .method(GET)
+                .uri("/check-tmp-file-still-exists");
+        result = route(app, request);
+        content = result.body().consumeData(mat).thenApply(bs -> bs.utf8String()).toCompletableFuture().get(5, TimeUnit.SECONDS);
+        assertEquals(OK, result.status());
+        assertEquals("not exists", content);
     }
 
     private void testTemporaryFile(final List<Http.MultipartFormData.FilePart> files) throws ExecutionException, InterruptedException, TimeoutException {

--- a/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/test/java/controllers/HomeControllerTest.java
+++ b/dev-mode/sbt-plugin/src/sbt-test/play-sbt-plugin/java-test-helpers/src/test/java/controllers/HomeControllerTest.java
@@ -88,6 +88,22 @@ public class HomeControllerTest extends WithApplication {
         testTemporaryFile(List.of(new Http.MultipartFormData.FilePart<>("document", "jabberwocky.txt", "text/plain", tempFile.path())));
     }
 
+    @Test
+    public void testTmpFileExists() throws IOException, ExecutionException, InterruptedException, TimeoutException {
+        play.api.libs.Files.TemporaryFile tempFile = play.api.libs.Files.SingletonTemporaryFileCreator$.MODULE$.create("temp", "txt");
+        write(tempFile.path(), "Hello".getBytes());
+
+        Http.RequestBuilder request = new Http.RequestBuilder()
+                .method(POST)
+                .bodyMultipart(Map.of(), List.of(new Http.MultipartFormData.FilePart<>("file", "file.txt", "text/plain", tempFile.path())))
+                .uri("/multipart-form-data-tmpfileexists");
+
+        Result result = route(app, request);
+        String content = result.body().consumeData(mat).thenApply(bs -> bs.utf8String()).toCompletableFuture().get(5, TimeUnit.SECONDS);
+        assertEquals(OK, result.status());
+        assertEquals("exists", content);
+    }
+
     private void testTemporaryFile(final List<Http.MultipartFormData.FilePart> files) throws ExecutionException, InterruptedException, TimeoutException {
         final Map<String, String[]> data = new HashMap<>();
         data.put("author", new String[]{"Lewis Carrol"});

